### PR TITLE
Alternative lisp modules

### LINKED ISSUE #142604: Common Lisp compilation files (fasl) are not being used
### Describe the bug
It seems that our shipped lisps are not using the package provided fasl files. I think the reason for this is a missing asdf-output translation.

### Steps To Reproduce
```sh
$ HOME=$(mktemp -d); cd

$ nix-shell -p lispPackages.cl-pdf --run "common-lisp.sh --eval '(require :cl-pdf)' --quit"
This is SBCL 2.1.2.nixos, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.

# lots of compilation output

$ ls -lah .cache/common-lisp/sbcl-2.1.2.nixos-linux-x64/nix/store/
total 0
drwxr-xr-x 6 user group 120 Oct 22 21:52 .
drwxr-xr-x 3 user group  60 Oct 22 21:52 ..
drwxr-xr-x 3 user group  60 Oct 22 21:52 hikw808j7b082z87d3rjmpz005m0bq5b-cl-wrapper-script
drwxr-xr-x 3 user group  60 Oct 22 21:52 hm0r08y0z9rz2yppi177j6qzdl3fl6sp-lisp-zpb-ttf-release-1.0.4
drwxr-xr-x 3 user group  60 Oct 22 21:52 llk9g31j4jzpmcy85vq747zwl61gwns0-lisp-iterate-20210228-git
drwxr-xr-x 3 user group  60 Oct 22 21:52 zkavzn6fh6v1bi2kmlw4382pwyfaws9z-lisp-cl-pdf-20210228-git

$
```

The lisp has recompiled the files into its cache in the user-directory, but the package ships with fasl files already:

```sh
$ ls /nix/store/zkavzn6fh6v1bi2kmlw4382pwyfaws9z-lisp-cl-pdf-20210228-git/lib/common-lisp/cl-pdf/*.fasl
/nix/store/zkavzn6fh6v1bi2kmlw4382pwyfaws9z-lisp-cl-pdf-20210228-git/lib/common-lisp/cl-pdf/bar-codes.fasl
/nix/store/zkavzn6fh6v1bi2kmlw4382pwyfaws9z-lisp-cl-pdf-20210228-git/lib/common-lisp/cl-pdf/chart.fasl
/nix/store/zkavzn6fh6v1bi2kmlw4382pwyfaws9z-lisp-cl-pdf-20210228-git/lib/common-lisp/cl-pdf/cl-pdf--system.fasl
# ...
$
```

I was able to get it somewhat to work by putting this content into `~/.sbclrc` or `~/.eclrc`:

```lisp
(require :asdf)
(asdf:initialize-output-translations '(:output-translations ("/nix/store" t) :inherit-configuration))
```

The problem with this is that now all derivations inside the nix-store need to ship with fasl files, otherwise we get something like a "read-only" error because the compiler tries to write its fasl file to the nix-store, which it cant. Ideally the directory `~/.cache/common-lisp/*/nix/store` should be empty or not exist at all.

I suggest that we properly setup the ASDF-output-translation to fix this, but I dont know how yet.

### Expected behavior
The lisp should use the fasl files provided by the derivation.

### Notify maintainers

CC @7c6f434c Do you know who else we can ping into the conversation?

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.14.12, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(user): `"emacs-overlay, home-manager, nur, nixos-21.05-21.05.1226.f77036342e2"`
 - channels(root): `"nixos-21.11pre323790.5f0194220f2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: lispPackages
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Uthar**
###### Description of changes

Alternative lisp-modules implementation. 

See discussion https://github.com/NixOS/nixpkgs/issues/155851

Summary of changes:

- Fixed loading of FASLs from the Nix store, which fixes #142604
- Faster way of importing Quicklisp releases - currently imports all systems directly from QL's `systems.txt`
- Added support for specifying Java libraries (ABCL)
- Added the `{abcl,ecl,ccl,sbcl}WithPackages` functions, which produce self-contained store closures, compared to the `common-lisp.sh` solution which is based on shell environments.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/lisp-modules/default.nix [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{ pkgs ? import <nixpkgs> { }, ... }:
+
+pkgs.callPackage ./nix-cl.nix { }
```
Comment (7c6f434c) [inline pkgs/development/lisp-modules/default.nix:3]: What this level of indirection is used for?

**Comment by 7c6f434c** [comment]
(I guess if this renamed the current `lispPackages` to `lispPackagesLegacy` and added the new version side by side as new `lispPackages`, it would be a more readable diff in GitHub… then we can dropping the legacy after a few weeks of trying things out can be discussed separately, and will have a pure-removal easy to review diff)

**Comment by Uthar** [comment]
> (I guess if this renamed the current `lispPackages` to `lispPackagesLegacy` and added the new version side by side as new `lispPackages`, it would be a more readable diff in GitHub… then we can dropping the legacy after a few weeks of trying things out can be discussed separately, and will have a pure-removal easy to review diff)

Thank you for the suggestion, I will fix this later this evening. In the meantime the second commit itself contains only additions, https://github.com/NixOS/nixpkgs/pull/170149/commits/e18582b985ceb0c509874941c969b97efbe1233d, so you could look there if you want.

**Comment by Uthar** [comment]
I added it as `pkgs.lispPackages_new` without touching the current code, this might be cleaner, do you agree?

File: pkgs/development/lisp-modules/default.nix (hunk 1 above)
Comment (Uthar) [reply pkgs/development/lisp-modules/default.nix]: it's unneeded, removed it now in https://github.com/NixOS/nixpkgs/commit/690b62640e38bbc1e9310b3532741fbd9c6cc58c

**Comment by Uthar** [comment]
Is `builtins.fetchTarball` supposed to be avoided? It seems to force some eager evaluation of the `src` attribute. The CI was complaining about it. `pkgs.fetchurl` doesn't have this problem.

**Comment by Uthar** [comment]
- Removed the asd conflict resolution code because it was too slow.
- Fixed eval time of big packages by removing slow code, so could remove the generic-cl hax from sqlite.lisp

**Comment by Uthar** [comment]
added a test to locally check if packages build

**Comment by Uthar** [comment]
switched to fetchzip to prevent downloading during evaluation

**Comment by Uthar** [comment]
pulled out the working makeWrapper from an older nixpkgs commit

File: pkgs/development/lisp-modules-new/doc/nix-cl.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+## Use cases
+
+This page lists some possible use cases for nix-cl.
+
+## Pinning down the exact commits of libraries
+
+Sometimes, a bug is fixed upstream but is not yet available in package
+repositories such as Quicklisp or Ultralisp. The users have to wait
```
Comment (7c6f434c) [inline pkgs/development/lisp-modules-new/doc/nix-cl.md:8]: I would probably say «a fix is available» if you want to mention Ultralisp — «fixed upstream» sounds like putting the fix into the main branch of the repo, and Ultralisp usually picks up things literally in a few minutes…

**Comment by 7c6f434c** [comment]
It's a pure addition, and I agree it is cleaner than the existing code, so let's merge it and see how much adapting the existing workflows need and based on that make a breaking-replacement warning/schedule later

**Comment by Uthar** [comment]
Awesome, yes let's see if we like it.

**Pull request merged**